### ISSUE TITLE
Access_Denied when implementing Microsoft Account Login & Issue with Lucene Indexer (Missing Segments file)

### ISSUE THREAD

**Issue by carnageX** [NuGet/NuGetGallery#3065]
Problem 1: 
I'm attempting to implement the NuGet gallery as a private repository in our organization.  Since it doesn't have LDAP auth in the current build (from what I saw only as a fork), I decided I would enable MS account login since everybody that would be using it has MSDN / MS accounts.  

Having an issue getting it working  however - I'm getting this as a response in the response headers after the MS login page pops up: /nugetgallery/users/account/authenticate/return?ReturnUrl=%2Fnugetgallery%2F&error=access_denied

Yes, I allowed the application to access my account.  

Problem 2: 
When I try to access the Packages URL on the site, I get this as an error: 
no segments* file found in Lucene.Net.Store.SimpleFSDirectory@<path to my application on a file share>\App_Data\Lucene lockFactory=Lucene.Net.Store.SingleInstanceLockFactory: files:

This happened when I tried moving to our Test environment.  It worked initially on our Dev environment.  When I copy all of the Lucene files that were generated from our Dev environment, it doesn't error out; but the indexer is wrong of course then, as there are no packages on our Test environment so the results are incorrect. 

Permissions to the App_Data folder are the same on our Test environment as they are on our Dev environment. 

Any help is appreciated, thanks!

**Comment by maartenba** [comment]
For 1, you may want to double check application settings at MS account. The error you are seeing comes from their end and usually is URL related. Not much we can help from our end I'm afraid.

For 2, clear the Lucene folder. The gallery will regenerate the index (you may want to restart the app pool to make this happen sooner)

**Comment by carnageX** [comment]
1.  It seems to be working initially - previously when I had an incorrect URL, it would just toss back an error mentioning that an error occurred.  Now it will at least prompt for the login and prompt to allow the gallery access to my account (if I haven't already given it access).  After that, it just tosses back the login page with "Your external account credential expired or you did not allow us access to your account. Try logging in again." displayed, but the Network tab of Chrome's debugger tools shows that the request to MS shows an error.  Not sure why this is happening exactly. 

2.  Unfortunately there's nothing to even clear out, as there's no files in it at all.  I'll try getting the app pool cleared tomorrow though. 

Thanks!

**Comment by maartenba** [comment]
Keep us posted! (closing for now)

**Comment by carnageX** [comment]
1.  The URL is correct according to documentation I've found about adding MS Account auth.  Here's a screenshot of the network hops:
![image](https://cloud.githubusercontent.com/assets/7111043/15860634/765e7a58-2c8f-11e6-9274-bcfca5a34387.png)

2. Unfortunately recycling the app pool didn't help; still receiving the "no segments* file found".

**Comment by maartenba** [comment]
Can you make sure the Lucene index folder as write access?

In terms of the MS Account settings, afraid I can't help there immediately as there are quite a few variables at play. Usually entering the full http(s) URL in the MS Account interface, and then configuring the Auth.MicrosoftAccount.ClientId, Auth.MicrosoftAccount.ClientSecret and Auth.MicrosoftAccount.Enabled should do the trick. Can you try entering just the root URL instead of the full path on MS Account side?

VS will not like self-signed certs, but a CA-signed SSL will work.

**Comment by carnageX** [comment]
Had our web team look at the permissions, and they said that it looked like the permissions were the same as our Dev environment (where it had initially worked).  How often does the Lucene index refresh/regenerate?  I tried moving the files on our Dev environment into a temp folder (to "remove" them), and then tried to access the Packages portion on Dev, and it blew up with the same "No segments* file" error.  Does the index not get generated on-the-fly when trying to access it when no Lucene files exist?  

Ok, looks like it was an issue on my end with not using the right ClientSecret - just me being dumb :).  Sorry about that! 

Looks like the Lucene issue is creeping into the API feed for VS as well, so not sure if that'll get fixed once I solve the Lucene issue or not.  If it doesn't, I will have our web team see if we can get a CA-assigned cert for the site.

**Comment by maartenba** [comment]
Can you try uploading a package? That triggers a rebuild

**Comment by carnageX** [comment]
Just uploaded a couple of packages and no change unfortunately.  

I can maybe have them check the event logs on the web server to see if it is indeed a permissions issue or if anymore light gets shed on the problem that way.

**Comment by carnageX** [comment]
Server logs didn't show anymore than what was already being displayed on screen with customErrors off.  

Here's the exception: 
Exception information: 
    Exception type: FileNotFoundException 
    Exception message: no segments* file found in Lucene.Net.Store.SimpleFSDirectory@<app_file_share_path>\App_Data\Lucene lockFactory=Lucene.Net.Store.SingleInstanceLockFactory: files:
   at Lucene.Net.Index.SegmentInfos.FindSegmentsFile.Run(IndexCommit commit) in d:\Lucene.Net\FullRepo\trunk\src\core\Index\SegmentInfos.cs:line 736
   at Lucene.Net.Index.DirectoryReader.Open(Directory directory, IndexDeletionPolicy deletionPolicy, IndexCommit commit, Boolean readOnly, Int32 termInfosIndexDivisor) in d:\Lucene.Net\FullRepo\trunk\src\core\Index\DirectoryReader.cs:line 117
   at Lucene.Net.Search.IndexSearcher..ctor(Directory path, Boolean readOnly) in d:\Lucene.Net\FullRepo\trunk\src\core\Search\IndexSearcher.cs:line 80
   at NuGetGallery.LuceneSearchService.SearchCore(SearchFilter searchFilter) in C:\Work\Projects\NuGetGallery\Source\Main\NuGetGallery\src\NuGetGallery\Infrastructure\Lucene\LuceneSearchService.cs:line 60
   at NuGetGallery.LuceneSearchService.Search(SearchFilter searchFilter) in C:\Work\Projects\NuGetGallery\Source\Main\NuGetGallery\src\NuGetGallery\Infrastructure\Lucene\LuceneSearchService.cs:line 50
   at NuGetGallery.PackagesController.<ListPackages>d__19.MoveNext() in C:\Work\Projects\NuGetGallery\Source\Main\NuGetGallery\src\NuGetGallery\Controllers\PackagesController.cs:line 406


Seems to be happening in LuceneSearchService.cs/SearchCore on line 60 when it tries to initialize the IndexSearcher.

**Comment by carnageX** [comment]
So I may have it fixed (?).  

Went into the Admin panel while logged into the site, and saw there was a "Lucene Maintenance" section.  Went there and clicked the "Rebuild Now" button, and it generated the necessary files and about 15-20min later it updated the Packages page.  Tested an upload...and it didn't update; waited for a couple of hours.  So I did another Rebuild Now and 20 min later it showed up.  So I went into the PackageController and ApiController and added the call to the "UpdateIndex" from the LuceneService, with a forceUpdate = true argument.  

Tested the package upload through both the browser and the API (separately of course), and each time after about 20min it showed up in the Packages page.  

So everything seems (?) to be working for now at least. Thanks for the help, appreciate it.
